db/revision_cache_lru.go

```diff
@@ -234,12 +255,28 @@ func (rc *LRURevisionCache) Put(ctx context.Context, docRev DocumentRevision) {
 		// TODO: CBG-1948
 		panic("Missing history for RevisionCache.Put")
 	}
-	value := rc.getValue(docRev.DocID, docRev.RevID, true)
+	var value *revCacheValue
+	if docRev.CV == nil {
+		value = rc.getValue(docRev.DocID, docRev.RevID, true)
+	} else {
+		value = rc.getValueByCV(docRev.DocID, docRev.RevID, docRev.CV, true)
+	}
+
 	value.store(docRev)
 }
 
 // Upsert a revision in the cache.
 func (rc *LRURevisionCache) Upsert(ctx context.Context, docRev DocumentRevision) {
+	// update the cv cache, then the revid cache too for backwards compatability
+	if docRev.CV != nil {
+		rc.upsertHLVCache(ctx, docRev)
+		// we have also updated the revID lookup cache too so return early here
+		return
+	}
+	rc.upsertRevIDCache(ctx, docRev)
```

Comment:
Can we stick to a single Upsert, and have the conditional handling around docRev.CV==nil inside the Upsert?